src/python/zquantum/core/hamiltonian.py

```diff
@@ -69,3 +71,124 @@ def group_comeasureable_terms_greedy(
             groups.append(QubitOperator(term, qubit_operator.terms[term]))
 
     return groups
+
+def compute_group_variances(groups: List[QubitOperator], expecval: ExpectationValues=None) -> np.array:
+    """Computes the variances of each frame in a grouped operator. 
+        If expectation values are provided, use variances from there, 
+        otherwise assume variances are 1 (upper bound). Correlation information
+        is ignored in the current implementation, covariances are assumed to be 0.
+    Args:
+        groups:  A list of QubitOperators that defines a (grouped) objective function
+        expecval: An ExpectationValues object containing the expectation
+                  values of the operators and their squares. Optionally, contains
+                  values of operator products to compute covariances.
+    Returns:
+        frame_variances: A Numpy array of the computed variances for each frame
+    """
+
+    if expecval is None:
+        frame_variances = [ np.sum(np.array(list(x.terms.values()))**2) for x in groups ] # Covariances are ignored; Variances are set to 1
+    else:
+        group_sizes = np.array([ len(g.terms.keys()) for g in groups ])
+        assert np.sum(group_sizes) == len(expecval.values)
+        real_expecval = expectation_values_to_real(expecval)
+        pauli_variances = 1. - real_expecval.values**2
+        frame_variances = []
+        for i, group  in enumerate(groups):
+            coeffs = np.array(list(group.terms.values()))
+            offset = 0 if i == 0 else np.sum(group_sizes[:i])
+            pauli_variances_for_group = pauli_variances[offset : offset + group_sizes[i]]
+            frame_variances.append(np.sum(coeffs**2 * pauli_variances_for_group))
+
+    return np.array(frame_variances)
+
+def get_expectation_values_from_rdms(interactionrdm: InteractionRDM, 
+        qubitoperator: QubitOperator, 
+        sort_terms: bool = False) -> ExpectationValues:
+    """ Computes expectation values of a qubitOperator
```

Comment:
The docstring here is inconsistent with previous docstrings in terms of indentation. It would be great if we could uniformize this.

Reply:
Sorry about that! I changed the indentation in the latest commit.